src/agent.py

```diff
@@ -96,4 +101,112 @@ def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
         event.relation.data[self.model.unit].update(
             AgentRelationData(url=f"http://{host}:{jenkins.WEB_PORT}", secret=secret)
         )
-        self.charm.unit.status = ActiveStatus()
+        self.charm.unit.status = ops.ActiveStatus()
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired from an agent joining the relationship.
+        """
+        container = self.charm.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
+            event.defer()
+            return
+        if not event.unit or not all(
+            event.relation.data[event.unit].get(required_data)
+            for required_data in ("executors", "labels", "name")
+        ):
+            logger.warning("Relation data not ready yet. Deferring.")
+            event.defer()
+            return
+        agent_meta = jenkins.AgentMeta(
+            executors=event.relation.data[event.unit]["executors"],
+            labels=event.relation.data[event.unit]["labels"],
+            name=event.relation.data[event.unit]["name"],
+        )
+        try:
+            agent_meta.validate()
+        except jenkins.ValidationError as exc:
+            logger.error("Invalid agent relation data. %s", exc)
+            self.charm.unit.status = ops.BlockedStatus("Invalid agent relation data.")
+            return
+
+        self.charm.unit.status = ops.MaintenanceStatus("Adding agent node.")
+        credentials = jenkins.get_admin_credentials(container)
+        try:
+            jenkins.add_agent_node(
```

Comment:
Couldn't this function just take the container and get its own credentials? Would reduce the coupling

Reply:
Refactored, thank you!